astropy/io/fits/hdu/table.py

```diff
@@ -206,31 +206,32 @@ def _init_tbdata(self, data):
         # new data placed in the column object above
         del columns._arrays
 
+    def _update_load_data(self):
+        """Load the data if asked to."""
+        if not self._data_loaded:
+            self.data
```

Comment:
The previous version was not working when the data is not loaded, and it's a bit tricky since calling `hdu.columns.add_col` modifies the `.columns` object which then makes it difficult to load correctly the data... So I used the notification mechanism to tell the HDU to load the data before columns are modified, which works and is quite simple though all of this is not really ideal.